src/Streamly/Internal/Data/Stream/StreamD.hs

```diff
@@ -1786,193 +1786,292 @@ data SplitOptions = SplitOptions
 
-        {-# INLINE go1 #-}
-        go1 !_ wrd st !acc = do
+    stepOuter gst (SplitOnSeqWordLoop wrd0 st0 fs0) =
+        go SPEC wrd0 st0 fs0
+
+        where
+
+        {-# INLINE go #-}
+        go !_ !wrd !st !fs = do
             res <- step (adaptState gst) st
             case res of
                 Yield x s -> do
-                    let wrd' = addToWord wrd x
-                        old = (mask .&. wrd) `shiftR` (elemBits * (patLen - 1))
-                    acc' <- fstep acc (toEnum $ fromIntegral old)
-                    if wrd' .&. mask == patWord
-                    then done acc' >>= \r -> return $ Yield r (GO_SHORT_PAT s)
-                    else go1 SPEC wrd' s acc'
-                Skip s -> go1 SPEC wrd s acc
-                Stop -> do
-                    acc' <- go2 wrd patLen acc
-                    done acc' >>= \r -> return $ Yield r GO_DONE
-
-        go2 !wrd !n !acc | n > 0 = do
-            let old = (mask .&. wrd) `shiftR` (elemBits * (n - 1))
-            fstep acc (toEnum $ fromIntegral old) >>= go2 wrd (n - 1)
-        go2 _ _ acc = return acc
+                    let wrd1 = addToWord wrd x
+                        old = (wordMask .&. wrd)
+                                `shiftR` (elemBits * (patLen - 1))
+                    fs1 <- fstep fs (toEnum $ fromIntegral old)
+                    if wrd1 .&. wordMask == wordPat
+                    then do
+                        r <- done fs1
+                        let next = SplitOnSeqWordInit s
+                        skip $ SplitOnSeqYield r next
+                    else go SPEC wrd1 s fs1
+                Skip s -> go SPEC wrd s fs
+                Stop -> skip $ SplitOnSeqWordDone patLen fs wrd
+
+    -------------------------------
+    -- General Pattern - Karp Rabin
+    -------------------------------
+
+    stepOuter gst (SplitOnSeqKRInit idx st rb rh) = do
+        res <- step (adaptState gst) st
+        case res of
+            Yield x s -> do
+                rh1 <- liftIO $ RB.unsafeInsert rb rh x
+                if idx == maxIndex
+                then do
+                    let fold = RB.unsafeFoldRing (RB.ringBound rb)
+                    let !ringHash = fold addCksum 0 rb
+                    if ringHash == patHash
+                    then do
+                        r <- initial >>= done
+                        let next = SplitOnSeqKRInit 0 s rb rh1
+                        -- XXX We need a direct yield here otherwise GHC is not
+                        -- able to fuse the code. This GHC issue needs to be
+                        -- investigated. We should know why.
+                        return $ Yield r next
+                    else do
+                        fs <- initial
+                        skip $ SplitOnSeqKRLoop fs s rb rh1 ringHash
+                else skip $ SplitOnSeqKRInit (idx + 1) s rb rh1
+            Skip s -> skip $ SplitOnSeqKRInit idx s rb rh
+            Stop -> do
```

Comment:
What is `idx == maxIndex` here and the pattern is matched?

Reply:
`idx` represents the next index that we are expecting. When `idx` == `maxIndex` and we got a `Stop` we have not got the last element, so it is not a match. It would be a match only in the `Yield` case.

Reply:
`What if`* was my question. Typo.

Reply:
Would be pretty stupid of me if I refactored `splitOnSeq` without knowing what `idx` was.